File: almoosa_customization/almoosa_customization/report/sales_summary_per_suppliers/sales_summary_per_suppliers.js
// Copyright (c) 2026, Printechs and contributors
// For license information, please see license.txt

frappe.query_reports["Sales Summary Per Suppliers"] = {
	"filters": [
		{
            "fieldname": "from_datetime",
            "label": "From Date & Time",
            "fieldtype": "Datetime",
            "default": frappe.datetime.month_start() + " 00:00:01"
        },

        {
            "fieldname": "to_datetime",
            "label": "To Date & Time",
            "fieldtype": "Datetime",
            "default": frappe.datetime.month_end() + " 23:59:59"
        },
		{
			"fieldname": "item_group",
			"label": "Item Group",
			"fieldtype": "MultiSelectList",
			"options": "Item Group",
			"get_data": function(txt) {
				return frappe.db.get_link_options("Item Group", txt).then(function(data) {
					let seen = new Set(); // to track duplicates
					let results = [];

					data.forEach(function(d) {
						let parts = d.value.split('.');
						let thirdPart = parts.length >= 3 ? parts[2] : d.value;

						if (!seen.has(thirdPart)) {
							seen.add(thirdPart);
							results.push({
								value: thirdPart,        // keep full name for backend
								description: thirdPart // 3rd part for display
							});
						}
					});

					return results;
				});
			}
		},



		{
			"fieldname": "warehouse",
			"label": "Warehouse",
			"fieldtype": "MultiSelectList",
			"options": "Warehouse",
			"get_data": function(txt) {
				return frappe.db.get_link_options("Warehouse", txt);
			}
		},
	],
	"formatter": function (value, row, column, data, default_formatter) {

        value = default_formatter(value, row, column, data);

        if (!data) {
            return value;
        }

        // ✅ Bold TOTAL column (all rows)
        if (column.fieldname === "total") {
            value = `<span style="font-weight:600;">${value}</span>`;
        }

        // ✅ Bold GRAND TOTAL row
        if (data.supplier === "Grand Total") {
            value = `<span style="font-weight:700;">${value}</span>`;
        }

        return value;
}
};

File: almoosa_customization/almoosa_customization/report/sales_summary_per_suppliers/sales_summary_per_suppliers.py
import frappe
from collections import defaultdict

def execute(filters=None):
    columns = []
    data = []

    conditions = ""
    values = {}

    # -----------------------
    # Date & Time Filter
    # -----------------------
    if filters.get("from_datetime") and filters.get("to_datetime"):
        conditions += """
            AND TIMESTAMP(hd.posting_date, hd.posting_time)
            BETWEEN %(from_datetime)s AND %(to_datetime)s
        """
        values["from_datetime"] = filters.get("from_datetime")
        values["to_datetime"] = filters.get("to_datetime")

    # -----------------------
    # Item Group Filter
    # -----------------------
    if filters.get("item_group"):
        conditions += """
            AND SUBSTRING_INDEX(SUBSTRING_INDEX(it.item_group, '.', 3), '.', -1) IN %(item_group)s
        """
        values["item_group"] = tuple(filters.get("item_group"))

    # -----------------------
    # Warehouse Filter
    # -----------------------
    if filters.get("warehouse"):
        conditions += " AND it.warehouse IN %(warehouse)s"
        values["warehouse"] = tuple(filters.get("warehouse"))

    # -----------------------
    # Fetch Data
    # -----------------------
    records = frappe.db.sql(f"""
        SELECT 
    MAX(sup.supplier) AS supplier,
    LEFT(it.warehouse,3) warehouse ,        
    SUM(it.qty) AS qty
FROM `tabPOS Invoice Item` AS it
INNER JOIN `tabPOS Invoice` hd 
    ON hd.name = it.parent
INNER JOIN `ViewSupplierItem` AS sup 
    ON sup.item_code = it.item_code
WHERE hd.docstatus = 1 {conditions}
GROUP BY sup.supplier,it.warehouse
ORDER BY sup.supplier;
    """, values, as_dict=True)

    # -----------------------
    # Prepare Pivot Structure
    # -----------------------
    supplier_data = defaultdict(lambda: defaultdict(float))
    warehouses = set()

    for row in records:
        supplier = row["supplier"]
        warehouse = row["warehouse"]
        qty = row["qty"] or 0

        supplier_data[supplier][warehouse] += qty
        warehouses.add(warehouse)

    warehouses = sorted(warehouses)

    # -----------------------
    # Columns
    # -----------------------
    columns.append({
        "label": "Supplier",
        "fieldname": "supplier",
        "fieldtype": "Data",
        "width": 90
    })

    for wh in warehouses:
        columns.append({
            "label": wh,
            "fieldname": wh,
            "fieldtype": "Int",
            "width": 60
        })

    columns.append({
        "label": "Total",
        "fieldname": "total",
        "fieldtype": "Int",
        "width": 100
    })

    # -----------------------
    # Data Rows
    # -----------------------
    grand_totals = defaultdict(float)

    for supplier in sorted(supplier_data.keys()):
        row = {"supplier": supplier}
        total = 0

        for wh in warehouses:
            qty = supplier_data[supplier].get(wh, 0)
            row[wh] = qty
            total += qty
            grand_totals[wh] += qty

        row["total"] = total
        data.append(row)

    # -----------------------
    # Grand Total Row
    # -----------------------
    if data:
        grand_row = {"supplier": "Grand Total"}
        grand_total_sum = 0

        for wh in warehouses:
            grand_row[wh] = grand_totals[wh]
            grand_total_sum += grand_totals[wh]

        grand_row["total"] = grand_total_sum
        data.append(grand_row)

    return columns, data
